snap/validate.go

```
@@ -433,50 +433,81 @@ func ValidatePathVariables(path string) error {
-	if li.Bind == "" && li.Type == "" && li.Symlink == "" {
-		return fmt.Errorf("cannot determine layout for %q", li.Path)
+	if layout.Symlink != "" {
+		nused += 1
 	}
-	if (li.Bind != "" && li.Type != "") ||
-		(li.Bind != "" && li.Symlink != "") ||
-		(li.Type != "" && li.Symlink != "") {
-		return fmt.Errorf("cannot accept conflicting layout for %q", li.Path)
+	if nused != 1 {
+		return fmt.Errorf("layout %q must define a bind mount, a filesystem mount or a symlink", layout.Path)
 	}
-	if li.Bind != "" {
-		if err := ValidatePathVariables(li.Bind); err != nil {
-			return fmt.Errorf("cannot accept layout of %q: %s", li.Path, err)
+
+	if layout.Bind != "" {
+		mountSource := layout.Bind
+		if err := ValidatePathVariables(mountSource); err != nil {
+			return fmt.Errorf("layout %q uses invalid bind mount source %q: %s", layout.Path, mountSource, err)
+		}
+		if !isAbsAndClean(mountSource) {
+			return fmt.Errorf("layout %q uses invalid bind mount source %q: must be absolute and clean", layout.Path, mountSource)
 		}
 	}
-	// Only the "tmpfs" filesystem is allowed.
-	if li.Type != "" && li.Type != "tmpfs" {
-		return fmt.Errorf("cannot accept filesystem %q for %q", li.Type, li.Path)
+
+	switch layout.Type {
+	case "tmpfs":
+	case "":
+		// nothing to do
+	default:
+		return fmt.Errorf("layout %q uses invalid filesystem %q", layout.Path, layout.Type)
 	}
-	if li.Symlink != "" {
-		if err := ValidatePathVariables(li.Symlink); err != nil {
-			return fmt.Errorf("cannot accept layout of %q: %s", li.Path, err)
+
+	if layout.Symlink != "" {
+		oldname := layout.Symlink
+		if err := ValidatePathVariables(oldname); err != nil {
+			return fmt.Errorf("layout %q uses invalid symlink old name %q: %s", layout.Path, oldname, err)
+		}
+		if !isAbsAndClean(oldname) {
+			return fmt.Errorf("layout %q uses invalid symlink old name %q: must be absolute and clean", layout.Path, oldname)
 		}
 	}
-	// Only certain users and groups are allowed.
+
+	switch layout.User {
```

Could you please add a comment here that any changes to the list of allowed user need to be reflected in spec.go? Or perhaps we should report internal error from there to make sure we do not silently go with a default of '0'.